salt/_modules/metalk8s_drain.py

```diff
@@ -0,0 +1,512 @@
+            pods = self.get_pods_for_eviction()
+        except DrainException as exc:
+            if dry_run:
+                pods = None
+            raise CommandExecutionError(
+                (
+                    "The following are not deletable: {0}. "
+                    "You can ignore DaemonSet pods with the "
+                    "ignore_daemonset flag."
+                ).format(
+                    exc.message
+                )
+            )
+
+        if dry_run:
+            return "Prepared for deletion of pods: {0}".format(
+                ", ".join(pods) if pods else "no pods to delete."
+            )
+
+        def get_pod(namespace, name):
+            '''Retrieve pod information'''
+            try:
+                api_instance = kubernetes.client.CoreV1Api()
+                api_response = api_instance.list_namespaced_pod(
+                    namespace=namespace,
+                    field_selector='metadata.name={0}'.format(name)
+                )
+                return api_response
+            except (ApiException, HTTPError) as exc:
+                if isinstance(exc, ApiException) and exc.status == 404:
+                    return None
+
+                log.exception(
+                    'Exception when calling '
+                    'CoreV1Api->list_namespaced_daemon_set')
+                raise CommandExecutionError(exc)
+
+        try:
+            self.evict_pods(pods, get_pod)
+        except DrainTimeoutException as exc:
+            raise CommandExecutionError(
+                "Drain did not complete within {0}".format(exc.message)
+            )
+        except DrainException as exc:
+            remaining_pods = self.get_pods_for_eviction()
+            raise CommandExecutionError(
+                "Pending pods remaining when the error occurred: {0}".format(
+                    exc),
+                [pod.metadata.name for pod in remaining_pods]
+            )
+        return "Eviction complete."
+
+    def evict_pods(self, pods, get_pod_fn):
+        '''Evict all pods from list'''
+        # TODO: there are a lot of shortcuts here, to be checked
+
+        global_timeout = self._timeout
+        for pod in pods:
+            # TODO: "too many requests" error not handled
+            _ = self.evict_pod(pod)
+
+        pending = self.wait_for_delete(
+            pods,
+            self.KUBECTL_INTERVAL,
+            global_timeout,
+            get_pod_fn
+        )
+
+        if pending:
+            raise DrainTimeoutException(global_timeout)
+
+    @staticmethod
+    def not_found(response):
+        '''Check whether a K8s API response indicates a resource not found.'''
+        return response is None or len(response.items) == 0
+
+    def wait_for_delete(self, pods, interval, timeout, get_pod_fn):
+        '''Wait for pods deletion.'''
+        total_t = 0
+        while total_t < timeout and pods:
+            pending = []
+            loop_s = time.time()    # For time processing pods
```

Comment:
```suggestion
            loop_s = time.time()  # For time processing pods
```

Oh and I don't really understand this variable name, but fine.